fix part of #19499: add prettier

nikitaevg opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #19499.
2. This PR does the following: 
    1. adds prettier .prettierrc.json, copied as-is from https://github.com/google/gts/blob/main/.prettierrc.json 
    1. adds lint-staged library and enables precommit formatting via lint-staged. Now every git commit will format the code via prettier
    1. disables eslint and tslint contradicting rules
    1. adds .prettierignore, which reduces impact of the change

The plan after this PR is merged is:
1. Update wiki about linters
1. Notify developers about new changes, because once the PR is merged the flow will be different.
1. format all the .ts and .js files
1. Figure out what to do with Github Action. It doesn't seem very trivial to set it up.


**Note:** Code formatted by prettier does not pass some custom eslint checks, for example custom [indentation](https://github.com/oppia/oppia/blob/a821335fee0ed436d0b2a231afcd6a78fb0f001a/.eslintrc#L240) and [break-after-parens](https://github.com/oppia/oppia/blob/a821335fee0ed436d0b2a231afcd6a78fb0f001a/.eslintrc#L158), probably some others. 

I suggest obeying prettier and removing all contradicting eslint rules, because it's Google config. _If this format works for Google it should work for us_. Because of this I'll first disable eslint for formatted files, format all the files and then enable eslint again.

This change brings two cool things:
1. Consistency - that's the main reason why I want this. The formatting is not strict at all, we can write code using different formats. With this change there will be single format of function definitions, function calls, if-s etc.
8. Much simpler formatting. Precommit will format code for us, no need to fix problems manually.

## Proof that changes are correct

N/A

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

nikitaevg committed:
fix part of #19499: add prettier

nikitaevg commented on .eslintrc:
```diff
@@ -4,7 +4,8 @@
   "extends": [
-    "eslint:recommended"
+    "eslint:recommended",
+    "prettier"
```
As per https://github.com/prettier/eslint-config-prettier?tab=readme-ov-file#installation

seanlip replied:
@nikitaevg Just a question about this. I thought we were planning to use a vanilla configuration, so wouldn't that entail getting rid of pretty much all the rules here? Or am I misunderstanding something?

nikitaevg replied:
These rules work well with prettier + I assume they were added for a reason, so why not keep them, that's my thinking. I could get rid of all of the rules here if  you want so, but the main thing I wanted to do with this change was to unify the style.

nikitaevg commented on .eslintrc:
```diff
@@ -155,7 +156,6 @@
     }
   ],
   "rules": {
-    "oppia/break-after-parens": "error",
```
the code formatted by prettier does not pass the removed rules.

nikitaevg commented:
Hi @seanlip, PTAL!

Please note the description in the PR description is meaningful:) Also please check the [state-responses.component.ts](https://github.com/oppia/oppia/pull/19854/files#diff-b1a08059405f09a5c2389211315839982e43497583b5a4fec2e4712ea57cfbd3) file as an example of what prettier does to files.

seanlip reviewed (commented):
@nikitaevg Heh, I always read the PR descriptions :P

I am a huge fan of this change. I support using the standard config for the reasons you mention. I would recommend going ahead with this (and let's also please add a step at the end to announce this to the devs too and explain how it affects their workflow). This will save lots of reviewer time, thank you for kicking it off!

seanlip commented:
P.S. I guess you plan to include the rest of the changes in this PR? I'm fine with that, though if you are planning to do this in stages it would be helpful to understand the PR-by-PR plan. (I haven't approved this for merge since it's still a draft, but I have no changes to suggest.)

nikitaevg commented:
> and let's also please add a step at the end to announce this to the devs too and explain how it affects their workflow

Sure, added

> I'm fine with that, though if you are planning to do this in stages it would be helpful to understand the PR-by-PR plan

I added my plan regarding distinct PRs into the description.

> I am a huge fan of this change.

Cool, thanks! I'll add the github action and the precommit hook then here and will send the PR for the review

nikitaevg commented on .prettierignore:
```diff
@@ -0,0 +1,18 @@
+core/templates/expressions/parser.js
```
Copied from .eslintignore + extra after local_compiled_js_for_test/*

nikitaevg committed:
update ignore

nikitaevg committed:
ignore json

nikitaevg committed:
remove whitespace problem; add yet another parsed file

nikitaevg committed:
add coverage exceptions

nikitaevg committed:
fix coverage exclusion

nikitaevg committed:
fix todo format

nikitaevg committed:
add prettier ignore

nikitaevg committed:
add backend incomplete coverage

nikitaevg committed:
add more ignore

nikitaevg committed:
fix todo

nikitaevg committed:
fix comment

nikitaevg committed:
remove extra newline

nikitaevg committed:
remove frontend coverage exception

nikitaevg commented:
@seanlip PTAL!

This is a preparation PR. I apply prettier to our codebase in [this](https://github.com/oppia/oppia/pull/19862) PR, which should be merged right after this one is merged.

After I merge both of these PRs I'll write the following email to the devs group (please tell me which one)

"""
Hi all!

I've just enabled the prettier formatter in our codebase which you can read more about on [this](https://github.com/oppia/oppia-web-developer-docs/blob/develop/Formatters.md) wiki page. If you have an open PR you will probably have merge conflicts. To resolve them merge the develop branch applying your changes and then run prettier. To do it run `git fetch -ap upstream && git merge upstream/develop --strategy-option ours && npx prettier . --write`. Drop me a message if you struggle after this change.

Thanks,
Nikita.
"""

seanlip reviewed (changes requested):
@nikitaevg Thanks, just one inline comment.

Also a general question -- the pre-commit hook can be bypassed. I think we need a lint check that verifies that running prettier on the code doesn't change it, or doesn't detect any violations, and it should be merged together with the 2nd PR so that we can guarantee that the codebase will follow prettier conventions from that point going forward. Can you add that since I'm planning to try and merge both PRs together?

Also speaking of which ... do both PRs need to be merged together? That is actually a bit tricky to do -- is it safe (i.e. won't break developers) to merge this one first and the other one a few days later? If it's not safe, what would break in that case?

Thanks!

nikitaevg reviewed (commented):
> Also a general question -- the pre-commit hook can be bypassed. I think we need a lint check that verifies that running prettier on the code doesn't change it, or doesn't detect any violations, and it should be merged together with the 2nd PR so that we can guarantee that the codebase will follow prettier conventions from that point going forward. Can you add that since I'm planning to try and merge both PRs together?

I thought about it as well, there's already a github action, but its setup is not straightforward to me. I thought about adding it once these two PRs are merged, but let me try again adding it in the second PR.

UPD: ok, it turned out to be much simpler, no external GitHub actions were needed. You can see the GitHub action in the [second PR](https://github.com/oppia/oppia/pull/19862), I updated the description accordingly.

> Also speaking of which ... do both PRs need to be merged together? That is actually a bit tricky to do -- is it safe (i.e. won't break developers) to merge this one first and the other one a few days later? If it's not safe, what would break in that case?

A few days later - not a great idea, because any commit will format all the changed files in that commit. As you said, one can ignore the pre-commit hook, but still, I'd rather these PRs were merged within hours from each other.

seanlip reviewed (approved):
Thanks! No particular concerns, this LGTM.
